File: app/services/api/src/dtos/input/group.py
from typing import Optional, List

from pydantic import BaseModel

from app.libs.db import models


class CreateGroup(BaseModel):
    """
    DTO for creating a new group
    """
    institution_id: str
    name: str
    parent_group_id: Optional[str] = None


class UpdateGroup(BaseModel):
    """
    DTO for updating an existing group
    """
    name: Optional[str] = None
    parent_group_id: Optional[str] = None


class UpdateGroupTimeslotPreferences(BaseModel):
    """
    DTO for updating timeslot preferences for a group (admin only)
    """
    preferences: List[models.TimeslotPreference]

File: app/services/ui/src/types/group.ts
import type { TimeslotPreference } from './user';

export interface GroupData {
  id?: string;
  _id?: string;
  institution_id: string;
  name: string;
  parent_group_id?: string | null;
  timeslot_preferences?: TimeslotPreference[];
}

export interface GroupShape {
  id: string;
  institution_id: string;
  name: string;
  parent_group_id?: string | null;
  timeslot_preferences: TimeslotPreference[];
}

export class Group implements GroupShape {
  id: string;
  institution_id: string;
  name: string;
  parent_group_id?: string | null;
  timeslot_preferences: TimeslotPreference[];

  constructor(data: GroupData) {
    this.id = data.id ?? data._id ?? '';
    this.institution_id = data.institution_id;
    this.name = data.name;
    this.parent_group_id = data.parent_group_id ?? null;
    this.timeslot_preferences = data.timeslot_preferences ?? [];
  }

  static from(data: GroupData): Group {
    return new Group(data);
  }
}
